### DISCUSSION TITLE
Affordance to dynamically determine which widget to call

### DISCUSSION THREAD

**Discussion by saqimtiaz** [TiddlyWiki/TiddlyWiki5#6491] [Ideas]
There are two related needs that keep coming up for me in different contexts:
1. an affordance to control which HTML element to create, for example `div` or `section`
2. an affordance to control which widget to invoke, for example `$edit-text` or `$edit-codemirror`

**Example Use Case 1**

A macro that accepts a `filter `and `tag` parameter and exports some text compiled from the tiddlers identified by the filter. The tag allows the user to control the structure of the export, for example whether to use `<div>` or `<section>` tags. Currently this requires using the wikify widget or nested macros and text substitution and is not very intuitive.

What would be more intuitive is something along these lines:
```
<$element tag=<<__tag__>> class="myclass">
<!--- content here --->
</$element>
```
_Whether we want to add user facing affordances for the `$element` widget or introduce a new widget needs consideration._

**Example Use Case 2**

A user configurable option that determines which editor engine to use, that is a flexible mechanism that controls which widget to invoke. Currently the only wikitext affordance for this is to wrap each widget in $list widgets, which does not scale very well.

Alternatively, something along these lines would be a lot more flexible, given a `configtiddler` variable where that tiddler is set to `codemirror` or `text`:
```
<$widgetcall $widget={{{ [<configtiddler>get[text]addprefix[edit-]] }}} focus="yes" tag="input" />
```

This would also open the door for deprecating the `$edit` widget in favour of a wikitext implementation, as  well as being useful in other contexts where there is a choice between similar or equivalent widgets, for instance one might be provided by a plugin and be more performant than the core alternative.

Potentially something along the lines of a `$widgetcall` widget could also be used for Example Use Case 1 above.

**Comment by Jermolene** [comment]
Hi @saqimtiaz those are both very interesting suggestions. I prefer the second approach. I am not keen on the namespace clash (ie that it wouldn't be possible to pass an attribute called `$widget`), but I think that boat has sailed.

**Reply by saqimtiaz** [reply]
Perhaps we could mitigate that to some extent by borrowing a trick from the `$action-log` widget and prefix attributes reserved for the `$widgetcall` widget itself with `$$` instead of `$`.

**Comment by saqimtiaz** [comment]
What follows are notes on my thinking process as I evaluate solutions. Next steps summarized at the bottom:

Example syntax:
`<$widgetcall $$widget=<<widget-type>>/>`

How to then differentiate between `button` and `$button`, i.e. use the same widget to render a widget or a DOM element?

---

Introduce optional `$$tag` attribute?

```
<$widgetcall $$widget="element" $$tag="div  class="myclass">

</$widgetcall>
```
outputs:
```
<div class="myclass">

</div>
```
---

Wikitext replacement for `$edit` widget:

```
\define editor(field:"text",tiddler)
<$let
	currentTiddler={{{ [<__tiddler__>!is[blank]else<currentTiddler>] }}}
	type={{{ [<__field__>match[text]then{!!type}!is[blank]] :else[[text/vnd.tiddlywiki]] }}}
	editor={{{ [<type>lookup[$:/config/EditorTypeMappings/]!is[blank]] :else[<currentTiddler>is[binary]then[binary]else[text]] :and[addprefix[$edit-]] }}}
	>
	<$widgetcall $$widget=<<editor>>/>
</$let>
\end

<$edit field="text" tiddler="MyTiddler">
<<edit field:"text" tiddler:"MyTiddler">>
```
---

**Next steps:**

Consider extending [html](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/parsers/wikiparser/rules/html.js) parser to support `$$element` as a user affordance for the `element` widget and `$$widget` as the equivalent for widgets.

Then both of the following would be possible:

```
<$$element $tag=<<tagName>> >
<!-- content here -->
</$$element>

<$$widget $$name=<<widgetName>> />
```

_Attribute names need further consideration in terms of consistency._
